### PR TITLE
PMM-4192 Updated MongoDB structs

### PR DESCRIPTION
**Pull request by percona-csalguero**
MongoDB explain was failing because Percona Toolkit is using the new MongoDB driver which uses a new bson.D type but it is not compatible with the old BsonD implemented by the toolkit. 
Now PT has been updated and this PR updates the dependency and fixes old data types.

### CODE REVIEW

**Comment by it-percona** [comment]
[CLA assistant check](https://cla-assistant.percona.com:443/percona/pmm-agent?pullRequest=174) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/pmm-agent?pullRequest=174) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: AlekSi<br/>:white_check_mark: BupycHuk<br/>:white_check_mark: percona-csalguero<br/>:x: cfsalguero<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/pmm-agent?pullRequest=174) it.</sub>

File: vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go [hunk 1]
```diff
@@ -120,21 +118,20 @@ func (f *Fingerprinter) Fingerprint(doc proto.SystemProfile) (Fingerprint, error
...
-					if f, ok := m["cond"]; ok {
-						if keysMap, ok := f.(bson.M); ok {
-							retKeys = append(retKeys, keys(keysMap, []string{})...)
-						}
-					}
-					if f, ok := m["ns"]; ok {
-						if ns, ok := f.(string); ok {
-							collection = ns
-						}
+				m, err := asMap(g)
```
Comment (BupycHuk) [inline vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go:121]: Previously we were ignoring if we can't cast

File: vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go [hunk 2]
```diff
@@ -271,3 +269,80 @@ func deduplicate(s []string) (r []string) {
...
+	// Proper way to detect if protocol used is "op_msg" or "op_command"
+	// would be to look at "doc.Protocol" field,
+	// however MongoDB 3.0 doesn't have that field
+	// so we need to detect protocol by looking at actual data.
+	query := doc.Query
+	if len(doc.Command) > 0 {
+		query = doc.Command
+		if doc.Op == "update" || doc.Op == "remove" {
+			if squery, ok := query.Map()["q"]; ok {
+				switch v := squery.(type) {
```
Comment (BupycHuk) [inline vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go:284]: can we use `asMap` here?

File: vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go (hunk 1 above)
Comment (percona-csalguero) [reply vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go:121]: Yes, but I think it was wrong. If the operation is `group`, there must be a `group` field.

File: vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go (hunk 2 above)
Comment (percona-csalguero) [reply vendor/github.com/percona/percona-toolkit/src/go/mongolib/fingerprinter/fingerprinter.go]: Done.

File: Gopkg.toml [hunk 3]
```diff
@@ -38,7 +38,8 @@ required = [
   version = "0.7.2"
 [[constraint]]
   name = "github.com/percona/percona-toolkit"
-  version = "3.2.0"
+  # version = "3.1.1"
+  branch = "PMM-4192_Mongo_explain_fails_fixing_tests"
```
Comment (askomorokhov) [inline Gopkg.toml:42]: 👀

**Comment by AlekSi** [comment]
I would say if we can do it there, we should. And looks like we can: https://github.com/percona/pmm-agent/blob/master/actions/mongodb_explain_action_test.go

File: agents/mongodb/internal/profiler/profiler_test.go [hunk 4]
```diff
@@ -185,6 +189,48 @@ func TestProfiler(t *testing.T) {
 	require.NotNil(t, findBucket)
 	assert.Equal(t, "FIND people name_00\ufffd", findBucket.Common.Fingerprint)
 	assert.Equal(t, docsCount, findBucket.Mongodb.MDocsReturnedSum)
+
```
Comment (percona-csalguero) [inline agents/mongodb/internal/profiler/profiler_test.go:192]: @AlekSi I added a new test here to ensure the queries the profiler gets are correct and we can use them for explain.

File: actions/mongodb_explain_action_test.go [hunk 5]
```diff
@@ -42,34 +46,60 @@ func TestMongoDBExplain(t *testing.T) {
...
+		assert.Nil(t, err)
+		queryPlanner, ok := explainM["queryPlanner"]
+		assert.Equal(t, ok, true)
+		assert.NotEmpty(t, queryPlanner)
+		assert.Equal(t, want, queryPlanner)
+	})
+
+	// PMM-4192 When we migrated
+	t.Run("Invalid MongoDB query", func(t *testing.T) {
+		fn := filepath.Join("/home/karl/go/src/github.com/percona/pmm-agent/testdata/mongo/", "doc-0161.bson")
```
Comment (AlekSi) [inline actions/mongodb_explain_action_test.go:82]: No such path on CI

File: actions/mongodb_explain_action.go [hunk 6]
```diff
@@ -62,12 +67,12 @@ func (a *mongodbExplainAction) Run(ctx context.Context) ([]byte, error) {
...
 	err = bson.UnmarshalExtJSON([]byte(a.params.Query), true, &eq)
 	if err != nil {
-		return nil, fmt.Errorf("explain: unable to decode query %s: %s", a.params.Query, err)
+		return nil, errors.Wrapf(err, "Query: %s", a.params.Query)
 	}
 
 	res := client.Database("admin").RunCommand(ctx, eq.ExplainCmd())
 	if res.Err() != nil {
-		return nil, res.Err()
+		return nil, errCannotExplain
```
Comment (AlekSi) [inline actions/mongodb_explain_action.go:75]: I think we should at least log the original error (`err` and `res.Err()`) there and above

File: actions/mongodb_explain_action_test.go [hunk 7]
```diff
@@ -42,34 +46,60 @@ func TestMongoDBExplain(t *testing.T) {
...
+			Dsn:   tests.GetTestMongoDBDSN(t),
+			Query: `{"ns":"test.coll","op":"query","query":{"k":{"$lte":{"$numberInt":"1"}}}}`,
+		}
 
-	ex := NewMongoDBExplainAction(id, params)
-	res, err := ex.Run(ctx)
-	assert.Nil(t, err)
+		ex := NewMongoDBExplainAction(id, params)
+		res, err := ex.Run(ctx)
+		assert.Nil(t, err)
```
Comment (AlekSi) [inline actions/mongodb_explain_action_test.go:57]: image

Is it `gofmt`'ed?

File: agents/mongodb/internal/profiler/profiler_test.go (hunk 4 above)
Comment (AlekSi) [reply agents/mongodb/internal/profiler/profiler_test.go:192]: 👍

File: actions/mongodb_explain_action_test.go (hunk 5 above)
Comment (percona-csalguero) [reply actions/mongodb_explain_action_test.go:82]: I am going to remove. This is a leftover I used to detect the error.

File: actions/mongodb_explain_action_test.go (hunk 7 above)
Comment (percona-csalguero) [reply actions/mongodb_explain_action_test.go:53]: Yes, I ran make format

Comment (AlekSi) [reply actions/mongodb_explain_action_test.go:53]: I wonder _how_ GitHub broke diff rendering just in that place

**Comment by AlekSi** [comment]
Is there a feature build?

File: Gopkg.toml (hunk 3 above)
Comment (AlekSi) [reply Gopkg.toml:42]: Please don't forget to update it

**Pull request merged**